src/h3lib/lib/h3Index.c

```diff
@@ -734,3 +760,232 @@ void H3_EXPORT(h3ToGeoBoundary)(H3Index h3, GeoBoundary* gb) {
+            if (testBaseCell == originBaseCell) {
+                break;
+            }
+        }
+        assert(revDir != 7);
+    }
+
+    int originOnPent = _isBaseCellPentagon(originBaseCell);
+    int indexOnPent = _isBaseCellPentagon(baseCell);
+
+    FaceIJK indexFijk = {0};
+    if (dir != 0) {
+        // Rotate index into the orientation of the origin base cell.
+        // cw because we are undoing the rotation into that base cell.
+        int baseCellRotations = baseCellNeighbor60CCWRots[originBaseCell][dir];
+        if (indexOnPent) {
+            for (int i = 0; i < baseCellRotations; i++) {
+                h3 = _h3RotatePent60cw(h3);
+
+                revDir = _rotate60cw(revDir);
+                if (revDir == 1) revDir = _rotate60cw(revDir);
+            }
+        } else {
+            for (int i = 0; i < baseCellRotations; i++) {
+                h3 = _h3Rotate60cw(h3);
+
+                revDir = _rotate60cw(revDir);
+            }
+        }
+    }
+    // Face is unused. This produces coordinates in base cell coordinate space.
+    _h3ToFaceIjkWithInitializedFijk(h3, &indexFijk);
+
+    // Origin leading digit -> index leading digit -> rotations 60 cw
+    // Either being 1 (K axis) is invalid.
+    // No good default at 0.
+    const int PENTAGON_ROTATIONS[7][7] = {
+        {0, -1, 0, 0, 0, 0, 0},        // 0
+        {-1, -1, -1, -1, -1, -1, -1},  // 1
+        {0, -1, 0, 0, 0, 1, 0},        // 2
+        {0, -1, 0, 0, 1, 1, 0},        // 3
+        {0, -1, 0, 5, 0, 0, 0},        // 4
+        {0, -1, 5, 5, 0, 0, 0},        // 5
+        {0, -1, 0, 0, 0, 0, 0},        // 6
+    };
+    // Simply prohibit many pentagon distortion cases rather than handling them.
+    const bool FAILED_DIRECTIONS_II[7][7] = {
+        {false, false, false, false, false, false, false},  // 0
+        {false, false, false, false, false, false, false},  // 1
+        {false, false, false, false, true, false, false},   // 2
+        {false, false, false, false, false, false, true},   // 3
+        {false, false, false, true, false, false, false},   // 4
+        {false, false, true, false, false, false, false},   // 5
+        {false, false, false, false, false, true, false},   // 6
+    };
+    const bool FAILED_DIRECTIONS_III[7][7] = {
+        {false, false, false, false, false, false, false},  // 0
+        {false, false, false, false, false, false, false},  // 1
+        {false, false, false, false, false, true, false},   // 2
+        {false, false, false, false, true, false, false},   // 3
+        {false, false, true, false, false, false, false},   // 4
+        {false, false, false, false, false, false, true},   // 5
+        {false, false, false, true, false, false, false},   // 6
+    };
+
+    if (dir != 0) {
+        assert(baseCell != originBaseCell);
+        assert(!(originOnPent && indexOnPent));
+
+        int pentagonRotations = 0;
+        int directionRotations = 0;
+
+        if (originOnPent) {
```

Comment:
I don't have a concrete suggestion here, but is it possible to move the pentagon handling blocks to a separate function to hide the complexity they entail?